### PR TITLE
Feature/2017 12 20 gauge blog

### PR DESCRIPTION
**Pull request by MartijnW49**

### CODE REVIEW

**Comment by MartijnW49** [comment]
Thanks @MrJean ! If ok, I'd like to wait untill friday to release it, since Tim will be releasing his post tomorrow and this one is kind of related.

**Comment by MrJean** [comment]
@MartijnW49 fine by me! Make sure to update your date to reflect the publishing date :)

File: _posts/2018-01-03-Gauge-automated-testing.md [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+
+After finishing a new feature in your web app, you test whether it works.
+However, while working on that new feature, you may have broken something else.
+If you want to know for sure, you have to test everything.
+Clicking around, filling in forms, &hellip; will cost you some valuable time.
+So, you'll probably end up not doing it thorough enough or even not at all, assuming everything else still works fine.
+That's exactly why you should have automated tests do it for you!
+
+[Gauge](https://getgauge.io/){:target="_blank"} is yet another test automation tool that serves that purpose.
+The founder is ThoughtWorks, the company that also created  <a href="http://www.seleniumhq.org/" target="_blank">Selenium</a> and <a href="https://www.gocd.org/" target="_blank">GoCD</a>.
```
Comment (yannickdeturck) [inline _posts/2018-01-03-Gauge-automated-testing.md]: You could make ThoughtWorks link to https://www.thoughtworks.com.

File: _posts/2018-01-03-Gauge-automated-testing.md [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+It's still in beta at the moment of writing, but it already works terrific!
+Gauge is comparable with Protractor or JUnit extended with Cucumber (if you haven't heard of these, they are worth checking out).
+In this blog we'll be mainly talking about automating browser testing using Selenium WebDriver, although that isn't the sole purpose of Gauge.
+If you don't know WebDriver, it's what allows you to interact with the browser to traverse the DOM, click on elements and so on.
+
+Now, unlike Cucumber, Gauge itself only supports three languages at the moment: Java, Ruby and C#.
+Others may be supported by the community like JavaScript.
+You can define steps written in either one of those languages.
+These are the actual tests and can be identified by a sentence you can choose yourself.
+By combining these sentences, you can write test scenarios which are written in MarkDown.
```
Comment (yannickdeturck) [inline _posts/2018-01-03-Gauge-automated-testing.md]: Markdown (lowercase 'd')

File: _posts/2018-01-03-Gauge-automated-testing.md [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+It can, for example, be used to install more plugins (as well as core plugins `gauge install java`).
+
+Next, in order to use WebDriver in our tests, we'll need the Java SDK and Maven.
+When you're a Java developer, you probably already have those installed.
+If not, you'll find enough on Google on how to install those.
+
+To start a new project, create a new folder and run `gauge init java` in it.
+This will setup a basic Gauge project.
+Then we need to add a [pom.xml](https://github.com/getgauge-examples/java-maven-selenium/blob/master/pom.xml){:target="_blank"} file ourselves, because we need some dependencies such as Selenium WebDriver.
+Next, WebDriver needs to be setup in our test code.
```
Comment (yannickdeturck) [inline _posts/2018-01-03-Gauge-automated-testing.md]: set up

File: _posts/2018-01-03-Gauge-automated-testing.md [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+The sentences can then be used to write the scenarios of our tests (or specs).
+If you'd like to combine multiple steps into once sentence, you can do so by creating so called "concepts".
+
+A typical folder structure for a Gauge setup using WebDriver and Java is as follows:
+<p>
+    <img class="image fit" style="max-width:194px" alt="Gauge Maven project structure" src="/img/2018-01-03-gauge-automated-testing/files.jpg" />
+</p>
+
+### Writing the specs
+The specs are written in MarkDown.
```
Comment (yannickdeturck) [inline _posts/2018-01-03-Gauge-automated-testing.md]: Markdown

File: _posts/2018-01-03-Gauge-automated-testing.md [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+Each spec file starts with a title and is underlined with `===`.
+Next, some steps can be defined that will be run before each scenario.
+When listing steps, you need to prefix each step with an asterix (*) as in a MarkDown list.
```
Comment (yannickdeturck) [inline _posts/2018-01-03-Gauge-automated-testing.md]: Markdown

File: _posts/2018-01-03-Gauge-automated-testing.md [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+After that, the actual scenarios can be written.
+They start with a title and are underlined with `---`.
+Again, the steps for a scenario should be listed as in a MarkDown list.
```
Comment (yannickdeturck) [inline _posts/2018-01-03-Gauge-automated-testing.md]: Markdown

File: _posts/2018-01-03-Gauge-automated-testing.md [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+Functions can be added to the page objects as well.
+Suppose you have a dropdown, you first have to click to open it and then select an option from the list.
+You can write a function doing all that.
+In your step definition, you then simply have to call that function to select something from a dropdown.
+It's a good way to avoid too much code duplication.
+
+### Concepts
+If you find yourself repeating the same sequence of steps over and over,
+you could combine those steps into one step using concepts.
+These are also written in MarkDown and you can pass arguments the same way as in the Java `@Step` annotations.
```
Comment (yannickdeturck) [inline _posts/2018-01-03-Gauge-automated-testing.md]: Markdown

File: _posts/2018-01-03-Gauge-automated-testing.md [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+
+`mvn gauge:execute -DspecsDir=specs -Dtags="sign-up & customer"`
+
+### Parallel
+So far, Gauge didn't stand out from other automated testing solutions.
+Although there's one thing that really does stand out and that's how easy it is to use parallel execution!
+
+`mvn gauge:execute -DspecsDir=specs -DinParallel=true`
+
+Running this command will look how many CPU cores your computer has and start up that many streams.
```
Comment (yannickdeturck) [inline _posts/2018-01-03-Gauge-automated-testing.md]: I think it sounds a bit better like this:
Running this command will start up a stream for each CPU core your computer has.

File: _posts/2018-01-03-Gauge-automated-testing.md [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+For each stream it will open a browser window and execute the specs.
+So if you have four CPU cores, four browser windows will be opened.
+You can overwrite the number of parallel executors, but it's recommended not to exceed the number of CPU cores.
+
+`mvn gauge:execute -DspecsDir=specs -DinParallel=true -Dnodes=2`
+
+Now, some specs may take longer to run than others.
+By default specs are divided dynamically over the streams.
```
Comment (yannickdeturck) [inline _posts/2018-01-03-Gauge-automated-testing.md]: By default, specs

File: _posts/2018-01-03-Gauge-automated-testing.md [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+So when a spec has finished, it will take the next one from the list of specs that still need to be executed.
+It's possible to change this so the specs are divided on startup, but the command is deprecated and will be removed.
+
+### Making tests independent
+To make use of this parallel execution, you'll have to make sure your tests don't rely on each other.
```
Comment (yannickdeturck) [inline _posts/2018-01-03-Gauge-automated-testing.md]: make sure **that** your

File: _posts/2018-01-03-Gauge-automated-testing.md [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+However, if you would want to test whether a user can sign in after having signed up, you should do so in one test.
+That immediately solves our problem and we are safe to use the parallel execution!
+
+# Environments and configuration
+The config files are located under `env/default`.
+You should have three files in that folder: `default.properties`, `java.properties` and `user.properties`.
+In the [example by Gauge](https://github.com/getgauge-examples/java-maven-selenium){:target="_blank"}, they have an `APP_URL` parameter in that last file.
+I recommend using that approach as well, you can get parameters in your Java code using `System.getenv("APP_URL")`.
+
+It's possible to create different environments by simply creating a new folder and add the configuration files that you want to change there.
```
Comment (yannickdeturck) [inline _posts/2018-01-03-Gauge-automated-testing.md]: ... and **adding** the ...
Not sure if I understand the last part of the sentence, "the configuration files that you want to change"?
You might mean specifying config files for multiple environment if certain properties differ depending on the environment?

File: _posts/2018-01-03-Gauge-automated-testing.md [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+If you need a different configuration for you CI-tool, you can create a folder named `ci`.
+When running the specs, you can pass an argument stating the environment.
+
+`mvn gauge:execute -Denv="ci"`
+
+Environments can also be used to run gauge with another browser like in the [example by Gauge](https://github.com/getgauge-examples/java-maven-selenium){:target="_blank"}.
+Create a folder named `firefox` for example and add a file called `browser.properties`.
+In that file you add `browser = FIREFOX`.
+When you then run the tests with the `firefox` environment, it will use FireFox as a browser instead.
+(This only works if you have your project setup like in the example, the [Driver](https://github.com/getgauge-examples/java-maven-selenium/tree/master/src/test/java/utils/driver){:target="_blank"} and [DriverFactory](https://github.com/getgauge-examples/java-maven-selenium/tree/master/src/test/java/utils/driver){:target="_blank"} files are required here.)
```
Comment (yannickdeturck) [inline _posts/2018-01-03-Gauge-automated-testing.md]: set up

**Comment by MartijnW49** [comment]
Thanks @yannickdeturck ! I fixed the nitpickings and clarified the part about the configuration. I also updated the date to tomorrow, so if you agree with the changes, you may merge this PR :)

**Pull request merged**